src/drawline.c

```diff
@@ -1314,11 +1314,18 @@ win_line(
 
 	    // When there was a sentence end in the previous line may require a
 	    // word starting with capital in this line.  In line 1 always check
-	    // the first word.
-	    if (lnum != capcol_lnum)
-		cap_col = -1;
-	    if (lnum == 1)
+	    // the first word.  Also check for sentence end in the line above
+	    // when updating changed text, the first row in a window, or if the
+	    // previous line is folded.
+	    if (lnum == 1
+		    || ((!nochange || startrow == 0
```

Comment:
So yeah just to be clear these two conditions could/should be combined into one that checks that we are in the first call to `win_line()` in a particular `win_update()`.
That would be the correct condition for when to check for sentence end in the previous line, as opposed to for all lines with edited text and the first row in a window.

Reply:
Changing the `nochange` argument to just be `mod_top` probably makes more sense:
```diff
diff --git a/src/drawline.c b/src/drawline.c
index c08aaa40d..3791ece43 100644
--- a/src/drawline.c
+++ b/src/drawline.c
@@ -1062,7 +1062,7 @@ win_line(
     linenr_T	lnum,
     int		startrow,
     int		endrow,
-    int		nochange UNUSED,	// not updating for changed text
+    int		mod_top UNUSED,	        // top line updated for changed text
     int		number_only)		// only update the number column
 {
     winlinevars_T	wlv;		// variables passed between functions
@@ -1316,10 +1316,9 @@ win_line(
 	    // word starting with capital in this line.  In line 1 always check
 	    // the first word.  Also check for sentence end in the line above
 	    // when updating the first row in a window, the top line with
-	    // changed text window, or if the previous line is folded.
+	    // changed text in a window, or if the previous line is folded.
 	    if (lnum == 1
-		    || ((startrow == 0
-			|| (!nochange && lnum == wp->w_buffer->b_mod_top)
+		    || ((startrow == 0 || mod_top == lnum
 #ifdef FEAT_FOLDING
 			|| hasFoldingWin(wp, lnum - 1, NULL, NULL, TRUE, NULL)
 #endif
@@ -2795,7 +2794,7 @@ win_line(
 			    p = prev_ptr;
 			cap_col -= (int)(prev_ptr - line);
 			len = spell_check(wp, p, &spell_hlf, &cap_col,
-								    nochange);
+								mod_top == 0);
 			word_end = v + len;
 
 			// In Insert mode only highlight a word that
diff --git a/src/drawscreen.c b/src/drawscreen.c
index a07c176d0..0912d7a15 100644
--- a/src/drawscreen.c
+++ b/src/drawscreen.c
@@ -2487,8 +2487,7 @@ win_update(win_T *wp)
 #endif
 
 		// Display one line.
-		row = win_line(wp, lnum, srow, wp->w_height,
-							  mod_top == 0, FALSE);
+		row = win_line(wp, lnum, srow, wp->w_height, mod_top, FALSE);
 
 #ifdef FEAT_FOLDING
 		wp->w_lines[idx].wl_folded = FALSE;
@@ -2535,7 +2534,7 @@ win_update(win_T *wp)
 		    fold_line(wp, fold_count, &win_foldinfo, lnum, row);
 		else
 #endif
-		    (void)win_line(wp, lnum, srow, wp->w_height, TRUE, TRUE);
+		    (void)win_line(wp, lnum, srow, wp->w_height, mod_top, TRUE);
 	    }
 
 	    // This line does not need to be drawn, advance to the next one.
```
I'll push one more time, feel free to include the previous version if you disagree.